src/session/key.ts

```diff
@@ -0,0 +1,126 @@
+/**
+ * SessionKey - Tracks the user's approval of a session key
+ *
+ * Session keys allow the user to authorize an app to take actions on
+ * their behalf without prompting their wallet for signatures.
+ * Session keys have a scope and an expiration.
+ * Session keys should be generated on the user's computer and persisted
+ * in a safe place or discarded.
+ *
+ * @example
+ * ```typescript
+ * const sessionKey = synapse.setSession(privateKey)
+ * await sessionKey.fetchExpiries()
+ * if (sessionKey.expiries[ADD_PIECES_TYPEHASH] * 1000 < Date.now() + HOUR_MILLIS) {
+ *   const DAY_MILLIS = 24 * HOUR_MILLIS
+ *   const loginTx = await sessionKey.login(BigInt(Date.now()) / BigInt(1000 + 30 * DAY_MILLIS), PDP_PERMISSIONS)
+ *   const loginReceipt = await loginTx.wait()
+ * }
+ * ```
+ */
+
+import { ethers } from 'ethers'
+import { getEIP712TypeString } from '../utils/eip712.ts'
+import { CONTRACT_ABIS, CONTRACT_ADDRESSES, getFilecoinNetworkType } from '../utils/index.ts'
+
+export const CREATE_DATA_SET_TYPEHASH = ethers.id(getEIP712TypeString('CreateDataSet'))
+export const ADD_PIECES_TYPEHASH = ethers.id(getEIP712TypeString('AddPieces'))
+export const SCHEDULE_PIECE_REMOVALS_TYPEHASH = ethers.id(getEIP712TypeString('SchedulePieceRemovals'))
+export const DELETE_DATA_SET_TYPEHASH = ethers.id(getEIP712TypeString('DeleteDataSet'))
+
+export const PDP_PERMISSIONS = [
+  CREATE_DATA_SET_TYPEHASH,
+  ADD_PIECES_TYPEHASH,
+  SCHEDULE_PIECE_REMOVALS_TYPEHASH,
+  DELETE_DATA_SET_TYPEHASH,
+]
+
+export const PDP_PERMISSION_NAMES: Record<string, string> = {
+  [CREATE_DATA_SET_TYPEHASH]: 'CreateDataSet',
+  [ADD_PIECES_TYPEHASH]: 'AddPieces',
+  [SCHEDULE_PIECE_REMOVALS_TYPEHASH]: 'SchedulePieceRemovals',
+  [DELETE_DATA_SET_TYPEHASH]: 'DeleteDataSet',
+}
+
+export class SessionKey {
+  private readonly _provider: ethers.Provider
+  private readonly _registry: ethers.Contract
+  private readonly _signer: ethers.Signer
+  private readonly _owner: ethers.Signer
+  public readonly expiries: Record<string, bigint | null>
+
+  public constructor(
+    provider: ethers.Provider,
+    sessionKeyRegistry: ethers.Contract,
+    signer: ethers.Signer,
+    owner: ethers.Signer
+  ) {
+    this._provider = provider
+    this._registry = sessionKeyRegistry.connect(owner) as ethers.Contract
```

Comment:
i think SessionKey should create and manage its own Contract instance and not rely on warmstorage class for this

Reply:
d1f8e5d